web-api/src/storage/elastic.rs

```diff
@@ -339,3 +331,181 @@ struct IngestedDocument<'a> {
+        params.embedding,
+        params.count,
+        params.num_candidates,
+        &inner_filter,
+    );
+
+    let mut generic_parameters = json_object!({
+        "size": params.count
+    });
+
+    if let Some(min_score) = params.min_similarity {
+        generic_parameters.extend(json_object!({
+            "min_score": min_score,
+        }));
+    }
+
+    KnnSearchParts {
+        knn_object,
+        generic_parameters,
+        inner_filter,
+    }
+}
+
+fn create_es_search_filter<'a>(
+    reference_time: DateTime<Utc>,
+    excluded_ids: impl IntoIterator<Item = &'a DocumentId>,
+    published_after: Option<DateTime<Utc>>,
+) -> JsonObject {
+    let time = reference_time.to_rfc3339();
+
+    // the existing documents are not filtered in the query to avoid too much work for a cold
+    // path, filtering them afterwards can occasionally lead to less than k results though
+    let excluded_ids = json!({
+        "ids": {
+            "values": excluded_ids.into_iter().collect_vec()
+        }
+    });
+
+    if let Some(published_after) = published_after {
+        // published_after != null && published_after <= publication_date <= time
+        json_object!({
+            "filter": {
+                "range": {
+                    "properties.publication_date": {
+                        "gte": published_after.to_rfc3339(),
+                        "lte": time
+                    }
+                }
+            },
+            "must_not": excluded_ids
+        })
+    } else {
+        // published_after == null || published_after <= time
+        json_object!({
+            "must_not": [
+                excluded_ids,
+                {
+                    "range": {
+                        "properties.publication_date": {
+                            "gt": time
+                        }
+                    }
+                }
+            ]
+        })
+    }
+}
+
+fn create_knn_request_object(
+    embedding: &NormalizedEmbedding,
+    count: usize,
+    num_candidates: usize,
+    filter: &JsonObject,
+) -> JsonObject {
+    // https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html
+    json_object!({
+        "knn": {
+            "field": "embedding",
+            "query_vector": embedding,
+            "k": count,
+            "num_candidates": num_candidates,
+            "filter": {
+                "bool": filter
+            }
+        }
+    })
+}
+
+fn normalize_scores<K>(mut scores: HashMap<K, f32>) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    let max_score = scores
+        .values()
+        .max_by(|l, r| l.total_cmp(r))
+        .copied()
+        .unwrap_or_default();
+    // TODO[pmk/now] the original code did a .max(1.0); this looked like bug, verify
+
+    for score in scores.values_mut() {
+        *score /= max_score;
+    }
+
+    scores
+}
+
+fn merge_scores_weighted<K>(
+    scores: impl IntoIterator<Item = (f32, HashMap<K, f32>)>,
```

Comment:
❓ should we enforce bounds on the weights (like each weight must be in `[0, 1]` and they must sum up to `1.`). maybe it's too much runtime effort to enforce it, but we should at least document it, otherwise the merged scores may not be normalized anymore.

Reply:
I don't think such constraints should be part of this function as it's not a constraint of this function but the environment we might call it in at best.